File: experimental/vibe/bench/python/latency.py
# SPDX-FileCopyrightText: Copyright (c) 2026, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
# SPDX-License-Identifier: Apache-2.0

"""Measure Python instrumentation-call latency with native pyperf timing."""

from __future__ import annotations

import logging
import os
from time import time_ns
from typing import Any

import pyperf
import quent_bench as quent
import structlog
from loguru import logger as loguru_logger
from opentelemetry.sdk._logs import LoggerProvider
from opentelemetry.sdk._logs.export import (
    ConsoleLogRecordExporter,
    LogRecordExporter,
    LogRecordExportResult,
    SimpleLogRecordProcessor,
)


class DiscardStructlogLogger:
    """Accept structured log calls without producing output."""

    def info(self, event: str, **kw: Any) -> None:
        del event, kw


class DiscardLogRecordExporter(LogRecordExporter):
    """Accept OpenTelemetry records synchronously without storing them."""

    def export(self, batch: Any) -> LogRecordExportResult:
        _ = len(batch)
        return LogRecordExportResult.SUCCESS

    def shutdown(self) -> None:
        pass

    def force_flush(self, timeout_millis: int = 10_000) -> bool:
        del timeout_millis
        return True


def capture_timestamp(
    logger: Any, method_name: str, event_dict: dict[str, Any]
) -> dict[str, Any]:
    """Capture an event timestamp without adding it as an attribute."""
    del logger, method_name
    _ = time_ns()
    return event_dict


def capture_timestamp_field(
    logger: Any, method_name: str, event_dict: dict[str, Any]
) -> dict[str, Any]:
    """Capture an event timestamp as structured output data."""
    del logger, method_name
    event_dict["timestamp_ns"] = time_ns()
    return event_dict


def export_path(filename: str) -> str:
    """Return a file path in the isolated output directory."""
    return os.path.join(os.environ["QUENT_BENCH_EXPORT_DIR"], filename)


def main() -> None:
    implementation = os.environ["QUENT_BENCH_IMPLEMENTATION"]
    output_mode = os.environ["QUENT_BENCH_OUTPUT_MODE"]
    if output_mode not in {"noop", "text"}:
        raise ValueError(f"unknown output mode: {output_mode}")
    prefix = f"latency/python/{output_mode}"
    runner = pyperf.Runner(processes=1, values=1, loops=1_000_000, warmups=0)
    runner.metadata["description"] = "Instrumentation call latency"
    if implementation == "clock":
        runner.timeit(f"{prefix}/clock", "time_ns()", globals={"time_ns": time_ns})
    elif implementation == "quent":
        context = (
            quent.Context()
            if output_mode == "noop"
            else quent.Context(quent.ExporterOptions.ndjson(export_path("quent")))
        )
        entity = context.entity_observer().handle()
        entity.tick()
        entity.tick()
        runner.timeit(f"{prefix}/quent", "entity.tick()", globals={"entity": entity})
        del entity
        context.close()
    elif implementation == "logging":
        entity_id = new_entity_id()
        logger = logging.getLogger("quent-bench")
        logger.handlers.clear()
        handler: logging.Handler
        if output_mode == "noop":
            handler = logging.NullHandler()
        else:
            handler = logging.FileHandler(export_path("logging.log"), mode="w")
            handler.setFormatter(
                logging.Formatter("%(asctime)s %(levelname)s entity_id=%(entity_id)s %(message)s")
            )
        logger.addHandler(handler)
        logger.propagate = False
        logger.setLevel(logging.INFO)
        bound_logger = logging.LoggerAdapter(logger, {"entity_id": entity_id})
        bound_logger.info("")
        runner.timeit(
            f"{prefix}/logging",
            'logger.info("")',
            globals={"logger": bound_logger},
        )
        handler.flush()
        handler.close()
        logger.handlers.clear()
    elif implementation == "structlog":
        output = None
        wrapped_logger: Any = DiscardStructlogLogger()
        processors = [capture_timestamp]
        if output_mode == "text":
            output = open(export_path("structlog.json"), "w", encoding="utf-8")
            wrapped_logger = structlog.PrintLogger(output)
            processors = [capture_timestamp_field, structlog.processors.JSONRenderer()]
        logger = structlog.wrap_logger(
            wrapped_logger,
            processors=processors,
            wrapper_class=structlog.stdlib.BoundLogger,
        ).bind(entity_id=new_entity_id())
        logger.info("")
        runner.timeit(
            f"{prefix}/structlog",
            'logger.info("")',
            globals={"logger": logger},
        )
        if output is not None:
            output.close()
    elif implementation == "loguru":
        loguru_logger.remove()
        sink = loguru_logger.add(
            (lambda message: None)
            if output_mode == "noop"
            else export_path("loguru.json"),
            level="INFO",
            format="{message}",
            serialize=output_mode == "text",
            enqueue=False,
            backtrace=False,
            diagnose=False,
        )
        logger = loguru_logger.bind(entity_id=new_entity_id())
        logger.info("")
        runner.timeit(
            f"{prefix}/loguru",
            'logger.info("")',
            globals={"logger": logger},
        )
        loguru_logger.remove(sink)
    elif implementation == "opentelemetry":
        output = None
        exporter: LogRecordExporter
        if output_mode == "noop":
            exporter = DiscardLogRecordExporter()
        else:
            output = open(export_path("opentelemetry.json"), "w", encoding="utf-8")
            exporter = ConsoleLogRecordExporter(out=output)
        provider = LoggerProvider(shutdown_on_exit=False)
        provider.add_log_record_processor(SimpleLogRecordProcessor(exporter))
        logger = provider.get_logger("quent-bench")
        attributes = {"entity.id": new_entity_id()}
        logger.emit(attributes=attributes)
        runner.timeit(
            f"{prefix}/opentelemetry",
            "logger.emit(attributes=attributes)",
            globals={"logger": logger, "attributes": attributes},
        )
        provider.shutdown()
        if output is not None:
            output.close()
    else:
        raise ValueError(f"unknown implementation: {implementation}")


def new_entity_id() -> str:
    """Create the entity UUID used by one isolated comparison process."""
    context = quent.Context()
    entity_id = str(context.entity_observer().handle().uuid)
    context.close()
    return entity_id


if __name__ == "__main__":
    main()

File: experimental/vibe/bench/cpp/latency.cpp
/*
 * SPDX-FileCopyrightText: Copyright (c) 2026, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
 * SPDX-License-Identifier: Apache-2.0
 */

#include <benchmark/benchmark.h>
#include <opentelemetry/exporters/ostream/log_record_exporter.h>
#include <opentelemetry/logs/log_record.h>
#include <opentelemetry/sdk/logs/exporter.h>
#include <opentelemetry/sdk/logs/logger_provider.h>
#include <opentelemetry/sdk/logs/logger_provider_factory.h>
#include <opentelemetry/sdk/logs/read_write_log_record.h>
#include <opentelemetry/sdk/logs/simple_log_record_processor_factory.h>
#include <quill/Backend.h>
#include <quill/Frontend.h>
#include <quill/LogMacros.h>
#include <quill/sinks/JsonSink.h>
#include <quill/sinks/NullSink.h>
#include <spdlog/logger.h>
#include <spdlog/sinks/basic_file_sink.h>
#include <spdlog/sinks/null_sink.h>

#include <chrono>
#include <cstdlib>
#include <filesystem>
#include <fstream>
#include <memory>
#include <stdexcept>
#include <string>
#include <utility>

#include "quent-bench-cpp-bridge/gen/quent.hpp"

namespace {

namespace otel_sdk_logs = opentelemetry::sdk::logs;
namespace otel_exporter_logs = opentelemetry::exporter::logs;

struct QuillFrontendOptions : quill::FrontendOptions {
  static constexpr quill::QueueType queue_type = quill::QueueType::BoundedBlocking;
  static constexpr std::size_t initial_queue_capacity = 1024u * 1024u;
};

using QuillFrontend = quill::FrontendImpl<QuillFrontendOptions>;
using QuillLogger = quill::LoggerImpl<QuillFrontendOptions>;

class DiscardLogExporter final : public otel_sdk_logs::LogRecordExporter {
 public:
  std::unique_ptr<otel_sdk_logs::Recordable> MakeRecordable() noexcept override {
    return std::make_unique<otel_sdk_logs::ReadWriteLogRecord>();
  }

  opentelemetry::sdk::common::ExportResult Export(
      const opentelemetry::nostd::span<std::unique_ptr<otel_sdk_logs::Recordable>>& records)
      noexcept override {
    benchmark::DoNotOptimize(records.size());
    return opentelemetry::sdk::common::ExportResult::kSuccess;
  }

  bool ForceFlush(std::chrono::microseconds) noexcept override { return true; }
  bool Shutdown(std::chrono::microseconds) noexcept override { return true; }
};

std::string new_entity_id() {
  auto context = quent::Context::none();
  auto entity = context.entity_observer()->handle();
  return quent::to_string(entity.id().raw());
}

std::string export_directory() {
  const char* path = std::getenv("QUENT_BENCH_EXPORT_DIR");
  if (path == nullptr || path[0] == '\0') {
    throw std::runtime_error("QUENT_BENCH_EXPORT_DIR must be set for Quent");
  }
  return path;
}

std::string export_path(const std::string& filename) {
  return (std::filesystem::path(export_directory()) / filename).string();
}

void Clock(benchmark::State& benchmark_state) {
  for (auto _ : benchmark_state) {
    benchmark::DoNotOptimize(std::chrono::system_clock::now());
  }
}

void benchmark_quent(benchmark::State& benchmark_state, bool text_output) {
  auto context = text_output ? quent::Context::ndjson(export_path("quent"))
                             : quent::Context::none();
  auto entity = context.entity_observer()->handle();
  entity.tick();
  entity.tick();
  for (auto _ : benchmark_state) {
    entity.tick();
    benchmark::ClobberMemory();
  }
}

void QuentNoop(benchmark::State& benchmark_state) { benchmark_quent(benchmark_state, false); }

void QuentText(benchmark::State& benchmark_state) { benchmark_quent(benchmark_state, true); }

void benchmark_spdlog(benchmark::State& benchmark_state, bool text_output) {
  const auto entity_id = new_entity_id();
  std::shared_ptr<spdlog::sinks::sink> sink;
  if (text_output) {
    sink = std::make_shared<spdlog::sinks::basic_file_sink_st>(export_path("spdlog.log"), true);
  } else {
    sink = std::make_shared<spdlog::sinks::null_sink_st>();
  }
  spdlog::logger logger("quent-bench", std::move(sink));
  logger.set_level(spdlog::level::trace);
  logger.info("entity_id={}", entity_id);
  for (auto _ : benchmark_state) {
    logger.info("entity_id={}", entity_id);
  }
  logger.flush();
}

void SpdlogNoop(benchmark::State& benchmark_state) { benchmark_spdlog(benchmark_state, false); }

void SpdlogText(benchmark::State& benchmark_state) { benchmark_spdlog(benchmark_state, true); }

void benchmark_quill(benchmark::State& benchmark_state, quill::ClockSourceType clock_source,
                     const std::string& logger_name, bool text_output) {
  const auto entity_id = new_entity_id();
  quill::Backend::start();
  std::shared_ptr<quill::Sink> sink;
  if (text_output) {
    sink = QuillFrontend::create_or_get_sink<quill::JsonFileSink>(
        export_path(logger_name + ".json"), quill::FileSinkConfig{});
  } else {
    sink = QuillFrontend::create_or_get_sink<quill::NullSink>("quent-bench-null");
  }
  auto* logger = QuillFrontend::create_or_get_logger(
      logger_name, std::move(sink), quill::PatternFormatterOptions{}, clock_source);
  QuillFrontend::preallocate();
  LOG_INFO(logger, "entity_id={}", entity_id);
  logger->flush_log();
  std::size_t pending = 0;
  for (auto _ : benchmark_state) {
    LOG_INFO(logger, "entity_id={}", entity_id);
    if (++pending == 4096) {
      benchmark_state.PauseTiming();
      logger->flush_log();
      pending = 0;
      benchmark_state.ResumeTiming();
    }
  }
  logger->flush_log();
  quill::Backend::stop();
}

void QuillTscNoop(benchmark::State& benchmark_state) {
  benchmark_quill(benchmark_state, quill::ClockSourceType::Tsc, "quent-bench-tsc", false);
}

void QuillTscText(benchmark::State& benchmark_state) {
  benchmark_quill(benchmark_state, quill::ClockSourceType::Tsc, "quent-bench-tsc", true);
}

void QuillSystemNoop(benchmark::State& benchmark_state) {
  benchmark_quill(benchmark_state, quill::ClockSourceType::System, "quent-bench-system", false);
}

void QuillSystemText(benchmark::State& benchmark_state) {
  benchmark_quill(benchmark_state, quill::ClockSourceType::System, "quent-bench-system", true);
}

void benchmark_opentelemetry(benchmark::State& benchmark_state, bool text_output) {
  const auto entity_id = new_entity_id();
  std::ofstream output;
  std::unique_ptr<otel_sdk_logs::LogRecordExporter> exporter;
  if (text_output) {
    output.open(export_path("opentelemetry.log"), std::ios::out | std::ios::trunc);
    exporter = std::make_unique<otel_exporter_logs::OStreamLogRecordExporter>(output);
  } else {
    exporter = std::make_unique<DiscardLogExporter>();
  }
  auto processor = otel_sdk_logs::SimpleLogRecordProcessorFactory::Create(
      std::move(exporter));
  auto provider = otel_sdk_logs::LoggerProviderFactory::Create(std::move(processor));
  auto logger = provider->GetLogger("quent-bench", "");
  const auto emit = [&] {
    auto record = logger->CreateLogRecord();
    record->SetAttribute("entity.id", entity_id);
    logger->EmitLogRecord(std::move(record));
  };
  emit();
  for (auto _ : benchmark_state) {
    emit();
  }
  provider->Shutdown();
}

void OpenTelemetryNoop(benchmark::State& benchmark_state) {
  benchmark_opentelemetry(benchmark_state, false);
}

void OpenTelemetryText(benchmark::State& benchmark_state) {
  benchmark_opentelemetry(benchmark_state, true);
}

BENCHMARK(Clock)->Name("latency/cpp/noop/clock");
BENCHMARK(QuentNoop)->Name("latency/cpp/noop/quent");
BENCHMARK(SpdlogNoop)->Name("latency/cpp/noop/spdlog");
BENCHMARK(QuillTscNoop)->Name("latency/cpp/noop/quill-tsc");
BENCHMARK(QuillSystemNoop)->Name("latency/cpp/noop/quill-system");
BENCHMARK(OpenTelemetryNoop)->Name("latency/cpp/noop/opentelemetry");
BENCHMARK(Clock)->Name("latency/cpp/text/clock");
BENCHMARK(QuentText)->Name("latency/cpp/text/quent");
BENCHMARK(SpdlogText)->Name("latency/cpp/text/spdlog");
BENCHMARK(QuillTscText)->Name("latency/cpp/text/quill-tsc");
BENCHMARK(QuillSystemText)->Name("latency/cpp/text/quill-system");
BENCHMARK(OpenTelemetryText)->Name("latency/cpp/text/opentelemetry");

}  // namespace

int main(int argc, char** argv) {
  benchmark::Initialize(&argc, argv);
  if (benchmark::ReportUnrecognizedArguments(argc, argv)) {
    return 1;
  }
  benchmark::RunSpecifiedBenchmarks();
  benchmark::Shutdown();
  return 0;
}
